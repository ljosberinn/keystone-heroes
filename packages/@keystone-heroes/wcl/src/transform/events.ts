import { EventType } from "@keystone-heroes/db/types";
import type { Prisma, Pull, PullZone } from "@keystone-heroes/db/types";
import type { DeepNonNullable } from "ts-essentials";

import { INVISIBILITY } from "../queries";
import { BOLSTERING } from "../queries/events/affixes/bolstering";
import { BURSTING } from "../queries/events/affixes/bursting";
import { EXPLOSIVE } from "../queries/events/affixes/explosive";
import { GRIEVOUS_WOUND } from "../queries/events/affixes/grievous";
import { NECROTIC } from "../queries/events/affixes/necrotic";
import { QUAKING } from "../queries/events/affixes/quaking";
import {
  SANGUINE_ICHOR_DAMAGE,
  SANGUINE_ICHOR_HEALING,
} from "../queries/events/affixes/sanguine";
import { STORMING } from "../queries/events/affixes/storming";
import { VOLCANIC } from "../queries/events/affixes/volcanic";
import { DOS_URN } from "../queries/events/dungeons/shadowlands/dos";
import { NW } from "../queries/events/dungeons/shadowlands/nw";
import {
  PF_GREEN_BUFF,
  PF_RED_BUFF,
  PF_PURPLE_BUFF,
} from "../queries/events/dungeons/shadowlands/pf";
import {
  SD_LANTERN_BUFF,
  SD_LANTERN_OPENING,
} from "../queries/events/dungeons/shadowlands/sd";
import { SOA_SPEAR } from "../queries/events/dungeons/shadowlands/soa";
import { TOP_BANNER_AURA } from "../queries/events/dungeons/shadowlands/top";
import type {
  AnyEvent,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
} from "../queries/events/types";
import type { ReportMapBoundingBox, ReportDungeonPullNpc } from "../types";

type Processor<T extends AnyEvent, AdditionalParams = Record<string, null>> = (
  event: T,
  params: {
    sourcePlayerID: number | null;
    targetPlayerID: number | null;
    sourceNPCID: number | null;
    targetNPCID: number | null;
  } & AdditionalParams
) => Prisma.EventCreateManyPullInput | null;

/**
 * track invis potion, bolstering, ToP and PF buffs
 */
const applyBuffProcessor: Processor<ApplyBuffEvent & { stacks?: number }> = (
  event,
  { sourcePlayerID, targetPlayerID, targetNPCID }
) => {
  if (sourcePlayerID && targetPlayerID) {
    if (
      sourcePlayerID === targetPlayerID &&
      (event.abilityGameID === INVISIBILITY.POTION_OF_THE_HIDDEN_SPIRIT ||
        event.abilityGameID === INVISIBILITY.DIMENSIONAL_SHIFTER)
    ) {
      return {
        timestamp: event.timestamp,
        sourcePlayerID,
        abilityID: event.abilityGameID,
        eventType: EventType.ApplyBuff,
      };
    }

    if (
      event.abilityGameID === TOP_BANNER_AURA ||
      event.abilityGameID === PF_GREEN_BUFF.aura ||
      event.abilityGameID === PF_RED_BUFF.aura ||
      event.abilityGameID === PF_PURPLE_BUFF.aura
    ) {
      return {
        timestamp: event.timestamp,
        eventType: EventType.ApplyBuff,
        abilityID: event.abilityGameID,
        sourcePlayerID,
        targetPlayerID,
      };
    }
  }

  if (event.abilityGameID === BOLSTERING && targetNPCID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuff,
      stacks: event.stacks ?? null,
      targetNPCID,
      targetNPCInstance: event.targetInstance ?? null,
    };
  }

  if (event.abilityGameID === SD_LANTERN_BUFF && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuff,
      targetPlayerID,
      abilityID: event.abilityGameID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const applyBuffStackProcessor: Processor<ApplyBuffStackEvent> = (
  event,
  { targetPlayerID }
) => {
  if (targetPlayerID && event.abilityGameID === SD_LANTERN_BUFF) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyBuffStack,
      abilityID: event.abilityGameID,
      targetPlayerID,
      stack: event.stack,
    };
  }

  return null;
};

/**
 * track SoA Spear & DOS Urn
 */
const applyDebuffProcessor: Processor<ApplyDebuffEvent> = (event) => {
  if (event.abilityGameID === SOA_SPEAR || event.abilityGameID === DOS_URN) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyDebuff,
      abilityID: event.abilityGameID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const beginCastProcessor: Processor<BeginCastEvent> = (
  event,
  { sourcePlayerID }
) => {
  if (sourcePlayerID && event.abilityGameID === SD_LANTERN_OPENING) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.BeginCast,
      abilityID: event.abilityGameID,
      sourcePlayerID,
    };
  }

  return null;
};

/**
 * track any cast from players
 * - pre-filtered to only contain 1min+ cooldowns
 * - dungeon-specific abilities (HoA Gargoyles)
 */
const castProcessor: Processor<CastEvent> = (
  event,
  { sourcePlayerID, targetNPCID, targetPlayerID }
) => {
  if (sourcePlayerID) {
    return {
      timestamp: event.timestamp,
      abilityID: event.abilityGameID,
      eventType: EventType.Cast,
      targetNPCID,
      targetPlayerID,
      sourcePlayerID,
    };
  }

  return null;
};

const isEnvironmentalDamageAffixAbilityID = (id: number) =>
  id === VOLCANIC ||
  id === STORMING ||
  id === EXPLOSIVE.ability ||
  id === GRIEVOUS_WOUND ||
  id === SANGUINE_ICHOR_DAMAGE ||
  id === BURSTING ||
  id === QUAKING ||
  id === NECROTIC;

/**
 * track any damage taken from
 * - affixes
 *
 * track any player damage done to
 * - explosives (pre-filtered to only contain lasthits)
 */
const damageProcessor: Processor<DamageEvent> = (
  event,
  { targetPlayerID, sourcePlayerID, sourceNPCID, targetNPCID }
) => {
  if (event.amount === 0) {
    return null;
  }

  // player taking damage
  if (targetPlayerID) {
    if (isEnvironmentalDamageAffixAbilityID(event.abilityGameID)) {
      return {
        timestamp: event.timestamp,
        abilityID: event.abilityGameID,
        eventType: EventType.DamageTaken,
        damage: event.amount + (event.absorbed ?? 0),
        targetPlayerID,
        sourcePlayerID,
      };
    }

    // NPC damaging player
    if (sourceNPCID) {
      return {
        timestamp: event.timestamp,
        eventType: EventType.DamageTaken,
        damage: event.amount + (event.absorbed ?? 0),
        targetPlayerID,
        sourceNPCID,
        // only relevant on plagueborer
        abilityID: event.abilityGameID,
      };
    }
  }

  // Player damaging NPC
  if (targetNPCID && sourcePlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.DamageDone,
      // damage vs Explosives is irrelevant
      damage:
        targetNPCID === EXPLOSIVE.unit
          ? null
          : event.amount + (event.absorbed ?? 0),
      targetNPCID,
      sourcePlayerID,
    };
  }

  return null;
};

/**
 * track deaths:
 * - all player deaths
 * - PF slimes
 */
const deathProcessor: Processor<DeathEvent, { pull: PersistedDungeonPull }> = (
  event,
  { targetNPCID, targetPlayerID, pull }
) => {
  const sourceNPCID = event.killerID
    ? pull.enemyNPCs.find((npc) => npc.id === event.killerID)?.gameID ?? null
    : null;

  // player death
  if (targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.Death,
      sourceNPCInstance: event.killerInstance ?? null,
      targetPlayerID,
      sourceNPCID,
    };
  }

  // prideful, PF slimes
  return {
    timestamp: event.timestamp,
    eventType: EventType.Death,
    targetNPCInstance: event.targetInstance ?? null,
    targetNPCID,
  };
};

/**
 * track any heal:
 * - sanguine
 * - kyrian orb
 */
const healProcessor: Processor<HealEvent> = (
  event,
  { targetNPCID, targetPlayerID, sourcePlayerID }
) => {
  if (event.amount === 0) {
    return null;
  }

  if (
    event.abilityGameID === NW.KYRIAN_ORB_HEAL &&
    sourcePlayerID &&
    targetPlayerID
  ) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.HealingDone,
      abilityID: NW.KYRIAN_ORB_HEAL,
      healingDone: event.amount,
      sourcePlayerID,
      targetPlayerID,
    };
  }

  if (event.abilityGameID === SANGUINE_ICHOR_HEALING && targetNPCID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.HealingDone,
      abilityID: SANGUINE_ICHOR_HEALING,
      healingDone: event.amount,
      targetNPCID,
    };
  }

  return null;
};

/**
 * track any interrupts through quaking including the interrupted ability
 */
const interruptProcessor: Processor<InterruptEvent> = (
  event,
  { sourcePlayerID, targetPlayerID }
) => {
  if (sourcePlayerID && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.Interrupt,
      interruptedAbilityID: event.extraAbilityGameID,
      sourcePlayerID,
      targetPlayerID,
    };
  }

  return null;
};

/**
 * track SD Lantern usage
 */
const removeBuffProcessor: Processor<RemoveBuffEvent> = (
  event,
  { targetPlayerID }
) => {
  if (event.abilityGameID === SD_LANTERN_BUFF && targetPlayerID) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.RemoveBuff,
      abilityID: event.abilityGameID,
      targetPlayerID,
    };
  }

  return null;
};

export type PersistedDungeonPull = Pick<
  Pull,
  "id" | "x" | "y" | "startTime" | "endTime" | "isWipe"
> & {
  maps: PullZone["zoneID"][];
  boundingBox: ReportMapBoundingBox;
  enemyNPCs: Pick<
    Required<DeepNonNullable<ReportDungeonPullNpc>>,
    "gameID" | "minimumInstanceID" | "maximumInstanceID" | "id"
  >[];
};

const getProcessorParams = (
  event: AnyEvent,
  pull: PersistedDungeonPull,
  actorPlayerMap: Map<number, number>
): Parameters<Processor<AnyEvent>>[1] => {
  const params: Parameters<Processor<AnyEvent>>[1] = {
    sourceNPCID: null,
    sourcePlayerID: null,
    targetNPCID: null,
    targetPlayerID: null,
  };

  if ("sourceID" in event) {
    params.sourcePlayerID = actorPlayerMap.get(event.sourceID) ?? null;
    params.sourceNPCID =
      pull.enemyNPCs.find((npc) => npc.id === event.sourceID)?.gameID ?? null;
  }

  if ("targetID" in event) {
    params.targetPlayerID = actorPlayerMap.get(event.targetID) ?? null;
    params.targetNPCID =
      pull.enemyNPCs.find((npc) => npc.id === event.targetID)?.gameID ?? null;
  }

  return params;
};

export const processEvents = (
  pull: PersistedDungeonPull,
  events: AnyEvent[],
  actorPlayerMap: Map<number, number>
): Omit<Prisma.EventCreateManyInput, "pullID">[] => {
  return events
    .map<Omit<Prisma.EventCreateManyInput, "pullID"> | null>((event) => {
      const params = getProcessorParams(event, pull, actorPlayerMap);

      switch (event.type) {
        case "applybuff":
          return applyBuffProcessor(event, params);
        case "applybuffstack":
          return applyBuffStackProcessor(event, params);
        case "applydebuff":
          return applyDebuffProcessor(event, params);
        case "begincast":
          return beginCastProcessor(event, params);
        case "cast":
          return castProcessor(event, params);
        case "damage":
          return damageProcessor(event, params);
        case "death":
          return deathProcessor(event, { ...params, pull });
        case "heal":
          return healProcessor(event, params);
        case "interrupt":
          return interruptProcessor(event, params);
        case "removebuff":
          return removeBuffProcessor(event, params);
        default:
          console.log(event);
          return null;
      }
    })
    .filter(
      (dataset): dataset is Omit<Prisma.EventCreateManyInput, "pullID"> =>
        dataset !== null
    );
};
