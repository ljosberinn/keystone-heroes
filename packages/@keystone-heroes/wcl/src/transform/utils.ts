import type { Prisma, Pull, PullZone } from "@keystone-heroes/db/types";
import type { DeepNonNullable } from "ts-essentials";

import type { AnyEvent } from "../queries/events/types";
import type { ReportMapBoundingBox, ReportDungeonPullNpc } from "../types";

export type Processor<
  T extends AnyEvent,
  AdditionalParams = Record<string, unknown>
> = (
  event: T,
  params: {
    sourcePlayerID: number | null;
    targetPlayerID: number | null;
    sourceNPCID: number | null;
    targetNPCID: number | null;
  } & AdditionalParams
) => Prisma.EventCreateManyPullInput | null;

export type PersistedDungeonPull = Pick<
  Pull,
  "id" | "x" | "y" | "startTime" | "endTime" | "isWipe" | "percent"
> & {
  maps: PullZone["zoneID"][];
  boundingBox: ReportMapBoundingBox;
  enemyNPCs: Pick<
    Required<DeepNonNullable<ReportDungeonPullNpc>>,
    "gameID" | "minimumInstanceID" | "maximumInstanceID" | "id"
  >[];
};
