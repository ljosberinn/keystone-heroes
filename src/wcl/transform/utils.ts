import type { Prisma, PullZone, Pull } from "@prisma/client";
import type { DeepNonNullable } from "ts-essentials";

import type { AnyEvent } from "../queries/events/types";
import type { ReportDungeonPullNpc } from "../types";

export type Processor<
  T extends AnyEvent,
  AdditionalParams = Record<string, unknown>
> = (
  event: T,
  params: {
    sourcePlayerID: number | null;
    targetPlayerID: number | null;
    sourceNPCID: number | null;
    targetNPCID: number | null;
  } & AdditionalParams
) => Prisma.EventCreateManyPullInput | null;

export type PersistedDungeonPull = Pick<
  Pull,
  "id" | "x" | "y" | "isWipe" | "percent"
> & {
  maps: PullZone["zoneID"][];
  enemyNPCs: Pick<
    Required<DeepNonNullable<ReportDungeonPullNpc>>,
    "gameID" | "id"
  >[];
  startTime: number;
  endTime: number;
};

export type PersistableDungeonPull = Omit<PersistedDungeonPull, "id"> & {
  count: number;
};
