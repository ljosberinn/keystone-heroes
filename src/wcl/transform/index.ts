import type { Prisma } from "@prisma/client";

import type { AllTrackedEventTypes } from "../queries/events/types";
import { applyBuffProcessor } from "./events/applybuff";
import { applyBuffStackProcessor } from "./events/applybuffstack";
import { applyDebuffProcessor } from "./events/applydebuff";
import { applyDebuffStackprocessor } from "./events/applydebuffstack";
import { beginCastProcessor } from "./events/begincast";
import { castProcessor } from "./events/cast";
import { damageProcessor } from "./events/damage";
import { deathProcessor } from "./events/death";
import { healProcessor } from "./events/heal";
import { interruptProcessor } from "./events/interrupt";
import { removeBuffProcessor } from "./events/removebuff";
import type { PersistedDungeonPull, Processor } from "./utils";

const getProcessorParams = (
  event: AllTrackedEventTypes,
  pull: PersistedDungeonPull,
  actorPlayerMap: Map<number, number>,
  allEnemyNPCs: PersistedDungeonPull["enemyNPCs"]
): Parameters<Processor<AllTrackedEventTypes>>[1] => {
  const params: Parameters<Processor<AllTrackedEventTypes>>[1] = {
    sourceNPCID: null,
    sourcePlayerID: null,
    targetNPCID: null,
    targetPlayerID: null,
  };

  if ("sourceID" in event) {
    params.sourcePlayerID = actorPlayerMap.get(event.sourceID) ?? null;
    params.sourceNPCID =
      pull.enemyNPCs.find((npc) => npc.id === event.sourceID)?.gameID ?? null;
  }

  if ("targetID" in event) {
    params.targetPlayerID = actorPlayerMap.get(event.targetID) ?? null;
    params.targetNPCID =
      allEnemyNPCs.find((npc) => npc.id === event.targetID)?.gameID ?? null;
  }

  return params;
};

export const processEvents = (
  pull: PersistedDungeonPull,
  events: AllTrackedEventTypes[],
  actorPlayerMap: Map<number, number>,
  allEnemyNPCs: PersistedDungeonPull["enemyNPCs"]
): Omit<Prisma.EventCreateManyInput, "pullID">[] => {
  return events
    .map<Omit<Prisma.EventCreateManyInput, "pullID"> | null>((event) => {
      const params = getProcessorParams(
        event,
        pull,
        actorPlayerMap,
        allEnemyNPCs
      );

      switch (event.type) {
        case "applybuff":
          return applyBuffProcessor(event, params);
        case "applydebuffstack":
          return applyDebuffStackprocessor(event, params);
        case "applybuffstack":
          return applyBuffStackProcessor(event, params);
        case "applydebuff":
          return applyDebuffProcessor(event, params);
        case "begincast":
          return beginCastProcessor(event, params);
        case "cast":
          return castProcessor(event, params);
        case "damage":
          return damageProcessor(event, params);
        case "death":
          return deathProcessor(event, { ...params, pull });
        case "heal":
          return healProcessor(event, params);
        case "interrupt":
          return interruptProcessor(event, params);
        case "removebuff":
          return removeBuffProcessor(event, params);
        default:
          return null;
      }
    })
    .filter(
      (dataset): dataset is Omit<Prisma.EventCreateManyInput, "pullID"> =>
        dataset !== null
    );
};
