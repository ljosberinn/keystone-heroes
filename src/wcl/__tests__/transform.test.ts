import { processEvents } from "../transform";
import allEvents from "./fixtures/allEvents.json";

describe("processEvents", () => {
  test("snapshot", () => {
    const pull: Parameters<typeof processEvents>[0] = {
      startTime: 0,
      endTime: 1,
      id: 1,
      isWipe: false,
      enemyNPCs: [],
      maps: [],
      x: 0,
      y: 0,
      percent: 0,
    };

    const result = processEvents(pull, allEvents, new Map(), []);

    expect(result).toMatchSnapshot();
  });
});
