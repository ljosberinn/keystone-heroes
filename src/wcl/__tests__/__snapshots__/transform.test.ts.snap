// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processEvents snapshot 1`] = `
[
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 647832,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4006933,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4018740,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4076733,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4078931,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4563925,
  },
  {
    "abilityID": 340225,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4563925,
  },
  {
    "abilityID": 340271,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4622916,
  },
  {
    "abilityID": 340271,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4978916,
  },
  {
    "abilityID": 340225,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4979920,
  },
  {
    "abilityID": 340225,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4986724,
  },
  {
    "abilityID": 340271,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4986724,
  },
  {
    "abilityID": 340378,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5032705,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5260722,
  },
  {
    "abilityID": 340210,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5262928,
  },
  {
    "abilityID": 340378,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5549610,
  },
  {
    "abilityID": 340378,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 6031023,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 11577910,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 11577910,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 11577910,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 11577910,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 11577910,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913879,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12269271,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12269271,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12269271,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12269271,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12269271,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12640578,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12640578,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12640578,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12640578,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 12640578,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 12840160,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 13482676,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 13482678,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 13482678,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 13482678,
  },
  {
    "abilityID": 340433,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 13482678,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 357900,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4094243,
  },
  {
    "abilityID": 357897,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4095304,
  },
  {
    "abilityID": 357575,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4096033,
  },
  {
    "abilityID": 357575,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4096613,
  },
  {
    "abilityID": 357575,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4098344,
  },
  {
    "abilityID": 357817,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4498851,
  },
  {
    "abilityID": 357814,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4500940,
  },
  {
    "abilityID": 357825,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4504750,
  },
  {
    "abilityID": 357815,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4506129,
  },
  {
    "abilityID": 357825,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4510612,
  },
  {
    "abilityID": 357706,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4741467,
  },
  {
    "abilityID": 357863,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4741845,
  },
  {
    "abilityID": 357747,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 4749497,
  },
  {
    "abilityID": 357897,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5090742,
  },
  {
    "abilityID": 356828,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5461614,
  },
  {
    "abilityID": 357524,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5463026,
  },
  {
    "abilityID": 357524,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5463652,
  },
  {
    "abilityID": 357556,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5465915,
  },
  {
    "abilityID": 356828,
    "eventType": "ApplyBuff",
    "targetPlayerID": null,
    "timestamp": 5466022,
  },
]
`;
